content/docs/iac/using-pulumi/extending-pulumi/build-a-provider.md

```diff
@@ -0,0 +1,756 @@
+
+{{% /notes %}}
+
+## Language support and the Pulumi Provider SDK
+
+Pulumi providers can be written in Go, TypeScript, .NET, and Java, and used in any Pulumi program, in any supported language. The [Pulumi Provider SDK](/docs/iac/using-pulumi/extending-pulumi/pulumi-provider-sdk/) is a framework for building providers in Go. We strongly recommend using the SDK, as it is the most full-featured and streamlined way to create a new provider.
+
+Some advantages of using the Pulumi Provider SDK:
+
+- **Minimal Code Required**: You define your resource types and implementation using Go structs and methods, and the SDK handles the rest (RPC, auto-generated schema for multi-language support, etc).
+- **Includes a Testing Framework**: Testing custom providers is made much easier with the SDK's built-in testing framework.
+- **Middleware Support**: Enhances providers with layers like token dispatch, schema generation, and cancellation propagation.
+
+## Example: Build a custom `file` provider
+
+Let's walk through the implementation of an example provider using the Pulumi Provider SDK. The [`file` provider](https://github.com/pulumi/pulumi-go-provider/blob/main/examples/file/main.go) will demonstrate how to manage local files as resources within Pulumi. It is a minimal but powerful illustration of the provider development process.
+
+### Features of the `file` provider
+
+The `file` provider can be used to create and modify a local file. Here's an example of how to use it in a Pulumi program:
+
+```yaml
+resources:
+  managedFile:
+    type: file:File
+    properties:
+      path: ${pulumi.cwd}/managed.txt
+      content: |
+        An important piece of information
+```
+
+In this example, we create a new resource called `managedFile` of type `file:File`. We can specify the path to write it to, and the contents that should be in it. During an update, Pulumi will use the `file` provider to ensure the file exists and has the specified contents.
+
+Now let's create the `file` provider that implements this.
+
+### Set up the project
+
+Since the provider is a separate code project from your Pulumi programs, the first step is to create a new directory for the provider code.
+
+```sh
+$ mkdir file-provider
+$ cd file-provider
+```
+
+{{< chooser language "go" >}}
+
+{{% choosable language go %}}
+
+#### Create `go.mod` file
+
+The `go.mod` file defined the Go project. It sets up the name of the Go module, the runtime requirements, and the module dependencies. We need the Pulumi Provider SDK (aka `github.com/pulumi/pulumi-go-provider`) as well as the standard Pulumi SDK (aka `github.com/pulumi/pulumi/sdk/v3`).
+
+```go
+module example.com/file-provider
+
+go 1.24
+
+toolchain go1.24.0
+
+require (
+	github.com/pulumi/pulumi-go-provider v1.0.0
+	github.com/pulumi/pulumi/sdk/v3 v3.167.0
+)
+```
+
+Once that file is created, go ahead and download the dependencies using `go get`:
+
+```sh
+$ go get example.com/file-provider
+```
```

Comment:
Do you mean "go mod tidy"?  Why would the user do `go get` on an unpublished thing?

Reply:
I'm not sure, but it works, and that's also the guidance I found when googling for the correct way to do this.